src/lib.rs

```diff
@@ -71,6 +55,41 @@ pub fn concat<'a>(a: &'a str, b: &'a str) -> Result<&'a str, Error> {
     }
 }
 
+fn concat_slice<'a>(a: &'a [u8], b: &'a [u8]) -> Result<&'a [u8], Error> {
+    let a_ptr = a.as_ptr();
+    let b_ptr = b.as_ptr();
+
+    let a_len = a.len();
+    let b_len = b.len();
+
+    // These should be guaranteed for the slices.
+    assert!(a_len <= isize::max_value() as usize);
+    assert!(b_len <= isize::max_value() as usize);
+
+    unsafe {
+        // https://doc.rust-lang.org/std/primitive.pointer.html#safety-1
+        // * starting pointer in-bounds obviously
+        // * ending pointer one byte past the end of an allocated object
+        // * explicit isize overflow check above
+        // * no wraparound required
+        // why: this is the one byte past the end pointer for the input slice `a`
+        if a_ptr.offset(a_len as isize) != b_ptr {
+            return Err(Error::NotAdjacent);
+        }
+        // UNWRAP: both smaller than isize, can't wrap in usize.
```

Comment:
I'd add an explanation here as to *why exactly* this is true in rust. For example in C the same could be UB, because in theory `ptrdiff_t` needs to be one bit larger than `uintptr_t`, because it's able to store any difference between two `uintptr_t`. This doesn't apply in rust due to the below reasons.

```suggestion
        // UNWRAP: both smaller than isize, can't wrap in usize.
        // This is because in rust `usize` and `isize` are both guaranteed to have
        // the same number of bits as a pointer [1]. As `isize` is signed, a `usize`
        // can always store the sum of two positive `isize`.
        // [1]: https://doc.rust-lang.org/reference/types/numeric.html#machine-dependent-integer-types
```